docs/architecture/adr-020-block-size.md

```diff
@@ -0,0 +1,66 @@
+# ADR 020: Limiting txs size inside a block
+
+## Changelog
+
+13-08-2018: Initial Draft
+15-08-2018: Second version after Dev's comments
+
+## Context
+
+We currently use MaxTxs to reap txs from the mempool when proposing a block,
+but enforce MaxBytes when unmarshalling a block, so we could easily propose a
+block thats too large to be valid.
+
+We should just remove MaxTxs all together and stick with MaxBytes, and have a
+`mempool.ReapMaxBytes`.
+
+But we can't just reap BlockSize.MaxBytes, since MaxBytes is for the entire block,
+not for the txs inside the block. There's extra amino overhead + the actual
+headers on top of the actual transactions + evidence + last commit.
+
+## Proposed solution
+
+Therefore, we should
+
+1) Get rid of MaxTxs.
+2) Rename MaxTxsBytes to MaxBytes.
+3) Add MaxHeaderBytes cs parameter (TODO: https://github.com/tendermint/tendermint/blob/8a1a79257e8e8b309cd35bb1fe40bf9b3330fd7d/types/block.go#L195)
```

Comment:
I'm confused as to why this is a consensus parameter.

AFAICT, everything here can be upper bounded from type definition:https://github.com/tendermint/tendermint/blob/8a1a79257e8e8b309cd35bb1fe40bf9b3330fd7d/types/block.go#L197-L222

All of the hashes are tmhashes, time.Time surely has a max size, we can bound chain-ID (256 bits b64 encoded? This is moreso dependent on what our plan for chain-ID is). We can use the max size for "int" in PartSetHeader (used within BlockID). So we should be able to derive the MaxHeaderBytes just from its definition. We can just amino encode it as well, so we don't have to guess at the amino overhead for the header.

Reply:
Right. Thanks.